packages/core/src/components/forms/numericStepper.tsx

```diff
@@ -0,0 +1,358 @@
+
+import { Button } from "../button/buttons";
+import { InputGroup } from "./inputGroup";
+
+export enum NumericStepperButtonPosition {
+    NONE,
+    LEFT,
+    RIGHT,
+    SPLIT,
+}
+
+export interface INumericStepperProps extends IIntentProps, IProps {
+
+    /**
+     * The button configuration with respect to the input field
+     * @default "right"
+     */
+    buttonPosition?: NumericStepperButtonPosition;
+
+    /**
+     * Whether the input is in a non-interactive state
+     * @default false
+     */
+    disabled?: boolean;
+
+    /** The ref handler that receives HTML `<input>` element backing this component */
+    inputRef?: (ref: HTMLInputElement) => any;
+
+    /** The name of icon (the part after `pt-icon-`) to render on left side of input */
+    leftIconName?: string;
+
+    /** The placeholder text in the absence of any value */
+    placeholder?: string;
+
+    /**
+     * The increment between successive values when `shift` is held
+     * @default 10
+     */
+    majorStepSize?: number;
+
+    /** The maximum value of the input */
+    max?: number;
+
+    /** The minimum value of the input */
+    min?: number;
+
+    /**
+     * The increment between successive values when `alt` is held
+     * @default 0.1
+     */
+    minorStepSize?: number;
+
+    /**
+     * The increment between successive values when no modifier keys are held
+     * @default 1
+     */
+    stepSize?: number;
+
+    /** The value to display in the input field */
+    value?: number | string;
+
+    /** The callback invoked when `enter` is pressed and when the field loses focus */
+    onDone?(value: string): void;
+
+    /** The callback invoked when the value changes */
+    onUpdate?(value: string): void;
+}
+
+export interface INumericStepperState {
+    shouldSelectAfterUpdate?: boolean;
+    value?: string;
+}
+
+@PureRender
+export class NumericStepper extends AbstractComponent<HTMLInputProps & INumericStepperProps, INumericStepperState> {
+    public static displayName = "Blueprint.NumericStepper";
+
+    public static defaultProps: INumericStepperProps = {
+        buttonPosition: NumericStepperButtonPosition.RIGHT,
+        majorStepSize: 10,
+        minorStepSize: 0.1,
+        stepSize: 1,
+        value: NumericStepper.VALUE_EMPTY,
+    };
+
+    private static DECREMENT_KEY = "decrement";
+    private static INCREMENT_KEY = "increment";
+    private static DECREMENT_ICON_NAME = "minus";
+    private static INCREMENT_ICON_NAME = "plus";
+    private static VALUE_EMPTY = "";
+    private static VALUE_ZERO = "0";
+
+    private inputElement: HTMLInputElement;
+
+    public constructor(props?: HTMLInputProps & INumericStepperProps) {
+        super(props);
+        this.state = {
+            shouldSelectAfterUpdate: false,
+            value: this.getValueOrEmptyValue(props),
+        };
+    }
+
+    public render() {
+        const { buttonPosition, className } = this.props;
+
+        const inputGroup = this.renderInputGroup();
```

Comment:
I think this code goes just a little beyond the point of producing optimal value from its DRYness. Storing this result produces extra indirection. Instead, this would read better if you inlined it in the 2 places it's used.